Add Basic Operations section with Alice and Bob flow. (#3)
bifurcation/ietf-mimi-protocol: draft-ralston-mimi-protocol.md

Terms and definitions are inherited from {{!I-D.barnes-mimi-arch}}.

> Throughout this document, the examples use the TLS Presentation Language
> {{!RFC8446}} and the semantics of HTTP {{!RFC7231}}. It is still an
> open issue what binary encoding the MIMI transport protocol will use,
> and there is at least one member of the design team who believes that
> HTTP semantics are not-appropriate.

Reviewer: ```suggestion
Throughout this document, the examples use the TLS Presentation Language
{{!RFC8446}} and the semantics of HTTP {{!RFC7231}}. 

> **TODO**: Decide on binary encoding and set of transport semantics.
```

# Basic Operations

This section walks a user through operations in the transport protocol in
the order they are necessary to show a basic messaging flow between Alice,
Bob, and Cathy:

> - Alice get the internal identifier for Bob †

Reviewer: This is out of scope for the design team's considerations, for now. We assume going into this document that an internal identifier has already been discovered through some means.

Later we may incorporate discovery, after the working group adopts a suitable document (or set of documents).

Author: Whether we address it or not, that step needs to happen, and an example flow needs to acknowledge that.

- Alice gains consent to talk to Bob ††

> - Alice fetches initial keying material for Bob's clients

Reviewer: This is just one possible flow: Alice can also create the room and *invite* Bob to it without ever requesting key material.

Reviewer 2: Just documenting one flow seems fine for this document, for now.

Reviewer: (we should at least mention that there's other possible flows though, if we still have them after our latest Alice and Bob discussions)

Author: We should document the most common flow *first* (Alice adds Bob), because with most messengers, if Alice "adds" Bob, she can send a message immediately which Bob can read after coming online (without any history passing mechanism).

- (Alice create a room) †††
- Alice adds Bob to a room
- Alice sends a message to the room
- Bob sends a message to the room

> - The hub/owning provider fans out the message

Reviewer: The participation changes are additionally fanned out, as is Alice's message. Which message is "the" message here?

Author: good point. it could be either.

- Bob adds a new client
- Alice repeats the first three steps for Cathy
- Alice adds Cathy to the room
- Alice removes Bob from the room

† is a placeholder for a later discovery mechanism.

†† there is no consensus in the design team about the full extent of

> consent requirements, but there is agreement that they will exist.

Reviewer: I don't think we need to highlight the design team's decisions like this in the doc. For consistency, let's use the `> **TODO**: Words` syntax.

††† as mentioned in the text, creating a room is a local provider
action, which is out of scope of MIMI.

> ## Get internal identifier for Bob

Reviewer: As discussed, this is out of scope for now. Let's remove it until we have an agreed working group item to work from.

mechanism. It is not intended to be directly implemented.

Alice fetches the internal identifier for some field of Bob's, in
this example his handle.

~~~
GET /identifierDiscovery/{domain}
~~~

The request body is described as:

~~~ tls
enum {
  reserved(0),
  handle(1),
  nick(2),
  email(3),
  phone(4),
  partialName(5),
  wholeProfile(6),
  oidcStdClaim(7),
  vcardField(8),
  (255)
} IdentifierType;

struct {
  IdentifierType type;
  string searchValue;
  select(type) {
     case oidcStdClaim:
       string claimName;
    case vcardField:
       string fieldName;
  };
} IdentifierRequest;

~~~

The response body is described as:

~~~ tls
enum {
  success(0),
  notFound(1),
  ambiguous(2),
  forbidden(3),
  unsupportedField(4),
  (255)
} IdentifierDiscoveryCode;

struct {
  IdentifierDiscoverCode responseCode;
  IdentifierUri uri;
} IdentifierResponse;
~~~

**TODO**: The format of specific identifiers is discussed in
{{?I-D.mahy-mimi-identity}}. Any specific conventions which are needed
should be merged into this document.

## Get consent for Bob

Alice sends a message requesting consent to add him to a room. At some
later time, Bob grants consent to Alice.

> Still **TODO** based on lack of consensus of consent scope.
> See Section 6.1 of {{?I-D.mahy-mimi-transport-design-reqs}} for
> some requirements.

Reviewer: ```suggestion
> **TODO**: Add mechanism to achieve this. See Section 6.1 
of {{?I-D.mahy-mimi-transport-design-reqs}} for thoughts.
```

> ## Fetch initial key material for Bob

Reviewer: This feels overly specific for this document. It's covered by MIMI-DS, and referenced by https://github.com/bifurcation/ietf-mimi-protocol/pull/4 per the agreed document layout.

This section should be removed.

} KeyMaterialRequest;
~~~

The response contains a list of users which contain a list of clients.
Each client provides keying material or returns a (client) error code.
Each user provides a list of clients (codes or material), and fully succeeds
(all clients succeed), partially succeed (some clients succeed), or returns
a (user) error code.

~~~ tls
enum {
    success(0);
    partialSuccess(1);
    incompatibleProtocol(2);
    noCompatibleMaterial(3);
    userUnknown(4);
    noConsent(5);
    noConsentForThisRoom(6);
    userDeleted(7);
    (255)
} KeyMaterialUserCode;

enum {
    success(0);
    keyMaterialExhausted(1);
    onlyLastResort(2);
    nothingComptible(3);
    (255)
} KeyMaterialClientCode;

struct {
    KeyMaterialClientCode clientStatus;
    IdentifierUri clientUri;
    select(protocol) {
        case mls10:
            KeyPackage keyPackage;
    };
} ClientKeyMaterial;

struct {
    KeyMaterialUserCode userStatus;
    IdentifiedUri userUri;
    ClientKeyMaterial clients<V>;
} UserKeyMaterial;

struct {
    Protocol protocol;
    UserKeyMaterial users<V>;
} KeyMaterialResponse;
~~~

**ISSUE**: Do we want key material requests to be combined for multiple users?

> ## Create a room

Reviewer: Room creation is covered elsewhere in the document. Can we add relevant detail to that instead, please?

Creating a room is done between a client and its local provider and is
out of scope of MIMI. However, we assume that the room has the following
policy document:

~~~
membershipStyle = membersOnly;
multiDevice = true;
moderated = false;
canAddUsers = [admin, owner];
canRemoveUsers = [admin, owner];

~~~

We also assume that Alice is the owner, and therefore an administrator,
of the room.

~~~
ownerRole = ["Alice@a.example"];
adminRole = ["Alice@a.example"];
activeParticipants = ["Alice@a.example"];
~~~

> ## Alice adds Bob

Reviewer: I'm not sure this needs to be a dedicated section, but much/all of the text contained within feels overly specified compared to the rest of the document. Can we merge relevant details into the other sections, please?

This comment applies to the section headers below as well.
